### PR TITLE
Support for AIToolBox, linear separable data creation

### PR DESCRIPTION
**Pull request by THaar50**
New changes include:
- Support for the AIToolBox framework with all algorithms that are supported on Linux
- Added support for creation of SEPARABLE as dataset which contains linear separable data 
- Added score handling functionality for Spark SVC models
- Bug fix for KSTest in sklearn tests to prevent NaNs as result values

### CODE REVIEW

File: src/main/java/atoml/data/DataDescription.java [hunk 1]
```diff
@@ -37,6 +37,11 @@
 	 * distribution used to generate numeric data
 	 */
 	private final AbstractRealDistribution distribution;
+	
+	/**
+	 * distribution used to generate numeric data for second class for separable data
+	 */
+	private final AbstractRealDistribution distribution2;
```
Comment (sherbold) [inline src/main/java/atoml/data/DataDescription.java]: please use a better name, e.g., distribution_separable or something similar that directly specifies the meaning of the attribute.

File: src/main/java/atoml/data/DataDescription.java [hunk 2]
```diff
@@ -174,6 +181,67 @@ public DataDescription(String name, int numFeatures, int numInformative, int num
...
+	 *            number of instances
+	 * @param distribution
+	 *            distribution of the numeric features
+	 * @param noiseRate
+	 *            noise rate
+	 * @param featureTypes
+	 *            types of the features; values>0 mean categorical features, the
+	 *            value defines the number of categories
+	 */
+	public DataDescription(String name, int numFeatures, int numInformative, int numInstances,
```
Comment (sherbold) [inline src/main/java/atoml/data/DataDescription.java:170]: Ideally try to have only one constructor with logic that is called by the others with default values. In this case, this should be this constructor that is called by the current constructors with distribution2=null. 

This removes code duplication and makes the class easier to maintain.

File: src/main/java/atoml/testgen/TestcaseGenerator.java [hunk 3]
```diff
@@ -64,7 +67,7 @@ public TestcaseGenerator(Algorithm algorithmUnderTest, List<DataDescription> mor
 		default:
 			// TODO
 			templateEngine = null;
-			System.err.format("unknown framework: %s", algorithmUnderTest.getFramework());
+			System.err.format("unknown framework: %s ", algorithmUnderTest.getFramework());
```
Comment (sherbold) [inline src/main/java/atoml/testgen/TestcaseGenerator.java]: undo?

File: src/main/java/atoml/testgen/TestcaseGenerator.java [hunk 4]
```diff
@@ -191,11 +200,14 @@ private String metamorphictestBody(MetamorphicTest metamorphicTest, DataDescript
...
 			replacementMap.put("<<<TIMEOUT>>>", System.getProperty("atoml." + this.algorithmUnderTest.getFramework() + ".timeout"));
 			replacementMap.put("<<<DATASET>>>", morphtestDataDescription.getName());
+			replacementMap.put("<<<CLASSNAME>>>", templateEngine.getClassName());
 			replacementMap.put("<<<ITERATIONS>>>", Integer.toString(iterations));
 			for (String key : replacementMap.keySet()) {
 				methodBody = methodBody.replaceAll(key, replacementMap.get(key));
 			}
+			
 			return methodBody;
+
```
Comment (sherbold) [inline src/main/java/atoml/testgen/TestcaseGenerator.java]: NIT: avoid new lines before closing a block, this is hard to read.

File: testdata/descriptions.yml [hunk 5]
```diff
@@ -2879,7 +2879,7 @@ parameters:
 # decision_function_shape, break_ties ignored because we only have binary problems so far    
 ---
 
-name: SKLEARN_LinearDiscriminantAnalysis_SVD
+name: SKLEARN_LinearDiscriminantAnalysisSVD
```
Comment (sherbold) [inline testdata/descriptions.yml]: please undo, the second underscores where intentional. Pattern: 
framework_algo_hyperparameterset

File: testdata/descriptions.yml [hunk 6]
```diff
@@ -2920,7 +2920,7 @@ parameters:
 # priors ignored
 ---
 
-name: SKLEARN_LinearDiscriminantAnalysis_LsqrEigen
+name: SKLEARN_LinearDiscriminantAnalysisLsqrEigen
```
Comment (sherbold) [inline testdata/descriptions.yml]: see above

File: testdata/descriptions.yml [hunk 7]
```diff
@@ -3232,11 +3232,15 @@ parameters:
...
-    default: None
+    default: 1.0
 #  affinity: # which affinity to use
 #    type: values
 #    values: [euclidean, precomputed]
 #    default: euclidean
+  random_state: # seed for starting state None=random
+    type: values
+    values: [0, None]
+    default: 0
```
Comment (sherbold) [inline testdata/descriptions.yml]: random state should not be part of the sampling and is handled by default with a fixed value

**Review by sherbold — COMMENTED** [review]
Looks mostly good, only some smaller issues. 

One additional thing: please add the limitations to the README, i.e., that mysql is not yet supported.

File: src/main/java/atoml/testgen/TestcaseGenerator.java (hunk 3 above)
Comment (THaar50) [reply src/main/java/atoml/testgen/TestcaseGenerator.java]: this was actually just a formatting fix, there was a space missing in the output when the error was thrown, should i revert?

Comment (sherbold) [reply src/main/java/atoml/testgen/TestcaseGenerator.java]: I am confused. The space you added is *after* the %s, shouldn't that just add a trailing space?

My guess is that the problem is different: format does not add a new line (like println), so maybe there should be a %n?

Comment (THaar50) [reply src/main/java/atoml/testgen/TestcaseGenerator.java]: Yes, exactly. It outputs the framework then a trailing space and follows it up with Exception in thread "main". A new line looks much better there!

**Comment by sherbold** [comment]
LGTM 👍

**Pull request merged**